src/snowflake/snowpark/_internal/ast.py

```diff
@@ -0,0 +1,94 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2012-2024 Snowflake Computing Inc. All rights reserved.
+#
+
+import base64
+import itertools
+import sys
+import uuid
+from dataclasses import dataclass
+from typing import Callable, Tuple
+
+import snowflake.snowpark._internal.proto.generated.ast_pb2 as proto
+from snowflake.snowpark.version import VERSION
+
+# The current AST version number (generated by the DSL).
+CLIENT_AST_VERSION = proto.__Version__.MAX_VERSION
+
+
+# TODO: currently unused.
+def expr_to_dataframe_expr(expr):
+    dfe = proto.SpDataframeExpr()
+    variant = expr.WhichOneof("variant")
+    getattr(dfe, variant).CopyFrom(getattr(expr, variant))
+    return dfe
+
+
+@dataclass
+class TrackedCallable:
+    var_id: int
+    func: Callable
+
+
+class AstBatch:
+    # Function used to generate request IDs. This is overridden in some tests.
+    generate_request_id = uuid.uuid4
+
+    def __init__(self, session) -> None:
+        self._session = session
```

Comment:
Don't want to bloat the scope of this PR unnecessarily, but I feel this PR may be a good place to add proper docstrings to all of the methods of AstBatch here. Some documentation could be useful for the team, especially for folks to understand assign/eval concept.